File: go/sdk/validation_test.go
package sdk

import (
	"strings"
	"testing"
)

func makeValidWasm(extra int) []byte {
	header := []byte{0x00, 0x61, 0x73, 0x6D, 0x01, 0x00, 0x00, 0x00}
	return append(header, make([]byte, extra)...)
}

func TestValidateBytecode_Valid(t *testing.T) {
	result := ValidateBytecode(makeValidWasm(100))
	if !result.Valid {
		t.Fatalf("expected valid, got errors: %v", result.Errors)
	}
	if len(result.Errors) != 0 {
		t.Fatalf("expected no errors, got %v", result.Errors)
	}
	if result.SizeBytes != 108 {
		t.Fatalf("expected 108 bytes, got %d", result.SizeBytes)
	}
}

func TestValidateBytecode_Empty(t *testing.T) {
	result := ValidateBytecode([]byte{})
	if result.Valid {
		t.Fatal("expected invalid for empty bytes")
	}
	if !strings.Contains(result.Errors[0], "empty") {
		t.Fatalf("expected 'empty' error, got %q", result.Errors[0])
	}
}

func TestValidateBytecode_TooSmall(t *testing.T) {
	result := ValidateBytecode([]byte{0x00, 0x61, 0x73})
	if result.Valid {
		t.Fatal("expected invalid for too-small bytes")
	}
	if !strings.Contains(result.Errors[0], "too small") {
		t.Fatalf("expected 'too small' error, got %q", result.Errors[0])
	}
}

func TestValidateBytecode_TooLarge(t *testing.T) {
	result := ValidateBytecode(makeValidWasm(512 * 1024))
	if result.Valid {
		t.Fatal("expected invalid for too-large bytes")
	}
	found := false
	for _, e := range result.Errors {
		if strings.Contains(e, "too large") {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected 'too large' error, got %v", result.Errors)
	}
}

func TestValidateBytecode_InvalidMagic(t *testing.T) {
	data := []byte{0xFF, 0xFF, 0xFF, 0xFF, 0x01, 0x00, 0x00, 0x00, 0x00}
	result := ValidateBytecode(data)
	if result.Valid {
		t.Fatal("expected invalid for bad magic")
	}
	found := false
	for _, e := range result.Errors {
		if strings.Contains(e, "magic") {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected 'magic' error, got %v", result.Errors)
	}
}

func TestValidateBytecode_InvalidVersion(t *testing.T) {
	data := []byte{0x00, 0x61, 0x73, 0x6D, 0x02, 0x00, 0x00, 0x00, 0x00}
	result := ValidateBytecode(data)
	if result.Valid {
		t.Fatal("expected invalid for bad version")
	}
	found := false
	for _, e := range result.Errors {
		if strings.Contains(e, "version") {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected 'version' error, got %v", result.Errors)
	}
}

func TestEstimateDeployGas_KnownSize(t *testing.T) {
	wasm := makeValidWasm(0) // 8 bytes
	expected := uint64(500_000 + 8*50)
	if got := EstimateDeployGas(wasm); got != expected {
		t.Fatalf("expected %d, got %d", expected, got)
	}
}

func TestEstimateDeployGas_ScalesWithSize(t *testing.T) {
	wasm := makeValidWasm(992) // 1000 bytes
	expected := uint64(500_000 + 1000*50)
	if got := EstimateDeployGas(wasm); got != expected {
		t.Fatalf("expected %d, got %d", expected, got)
	}
}

File: typescript/test/validation.test.mjs
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { validateBytecode, estimateDeployGas } from "../dist/validation.js";

/** Build a valid WASM header with optional extra padding bytes. */
function makeValidWasm(extra = 0) {
  const header = new Uint8Array([0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00]);
  const buf = new Uint8Array(8 + extra);
  buf.set(header);
  return buf;
}

describe("validateBytecode", () => {
  it("accepts valid WASM", () => {
    const result = validateBytecode(makeValidWasm(100));
    assert.equal(result.valid, true);
    assert.deepEqual(result.errors, []);
    assert.equal(result.sizeBytes, 108);
  });

  it("rejects empty bytes", () => {
    const result = validateBytecode(new Uint8Array(0));
    assert.equal(result.valid, false);
    assert.ok(result.errors[0].includes("empty"));
    assert.equal(result.sizeBytes, 0);
  });

  it("rejects too-small bytes", () => {
    const result = validateBytecode(new Uint8Array([0x00, 0x61, 0x73]));
    assert.equal(result.valid, false);
    assert.ok(result.errors[0].includes("too small"));
  });

  it("rejects too-large bytes", () => {
    const result = validateBytecode(makeValidWasm(512 * 1024));
    assert.equal(result.valid, false);
    assert.ok(result.errors.some((e) => e.includes("too large")));
  });

  it("rejects invalid magic bytes", () => {
    const bytes = new Uint8Array([0xff, 0xff, 0xff, 0xff, 0x01, 0x00, 0x00, 0x00, 0x00]);
    const result = validateBytecode(bytes);
    assert.equal(result.valid, false);
    assert.ok(result.errors.some((e) => e.includes("magic")));
  });

  it("rejects invalid version", () => {
    const bytes = new Uint8Array([0x00, 0x61, 0x73, 0x6d, 0x02, 0x00, 0x00, 0x00, 0x00]);
    const result = validateBytecode(bytes);
    assert.equal(result.valid, false);
    assert.ok(result.errors.some((e) => e.includes("version")));
  });
});

describe("estimateDeployGas", () => {
  it("returns correct estimate for known size", () => {
    const wasm = makeValidWasm(0); // 8 bytes
    assert.equal(estimateDeployGas(wasm), 500_000n + 8n * 50n);
  });

  it("scales with bytecode size", () => {
    const wasm = makeValidWasm(992); // 1000 bytes
    assert.equal(estimateDeployGas(wasm), 500_000n + 1000n * 50n);
  });
});
